src/human_encoding/mod.rs

```diff
@@ -223,76 +223,108 @@ impl<J: Jet> Forest<J> {
+            .finalize()
+        {
+            Ok(program) => program,
+            Err(error) => {
+                assert_eq!(error.to_string().as_str(), err_msg);
+                return;
+            }
+        };
+        let mut mac = BitMachine::for_program(&program);
+        match mac.exec(&program, env) {
+            Ok(_) => panic!("Execution is expected to fail"),
+            Err(error) => assert_eq!(error.to_string().as_str(), err_msg),
         }
     }
 
     #[test]
-    fn to_witness_node_unfilled_hole() {
-        let s = "
-            wit1 := witness
-            main := comp wit1 comp disconnect iden ?dis2 unit
-        ";
-        let forest = Forest::<Core>::parse(s).unwrap();
-        let witness = HashMap::new();
-
-        match forest.to_witness_node(&witness).unwrap().finalize() {
-            Ok(_) => panic!("Duplicate witness names should fail"),
-            Err(Error::IncompleteFinalization) => {}
-            Err(error) => panic!("Unexpected error {}", error),
-        }
+    fn unfilled_witness() {
+        let witness = HashMap::from([(Arc::from("wit1"), Value::u32(1337))]);
+        assert_finalize_err::<Core>(
+            "
+                wit1 := witness : 1 -> 2^32
+                wit2 := witness : 1 -> 2^32
+
+                wits_are_equal := comp (pair wit1 wit2) jet_eq_32 : 1 -> 2
+                main := comp wits_are_equal jet_verify            : 1 -> 1
+            ",
+            &witness,
+            &(),
+            "unable to satisfy program",
+        );
     }
 
     #[test]
-    fn to_witness_node_pruned_witness() {
+    fn unfilled_witness_pruned() {
         let s = "
             wit1 := witness
             wit2 := witness
             main := comp (pair wit1 unit) case unit wit2
         ";
-        let forest = Forest::<Core>::parse(s).unwrap();
-        let mut witness = HashMap::new();
-        witness.insert(Arc::from("wit1"), Value::u1(0));
-
-        match forest.to_witness_node(&witness).unwrap().finalize() {
-            Ok(_) => {}
-            Err(e) => panic!("Unexpected error {}", e),
-        }
-
-        witness.insert(Arc::from("wit1"), Value::u1(1));
-
-        match forest.to_witness_node(&witness).unwrap().finalize() {
-            Ok(_) => {}
-            Err(Error::IncompleteFinalization) => {}
-            Err(e) => panic!("Unexpected error {}", e),
-        }
-
-        witness.insert(Arc::from("wit2"), Value::unit());
+        let wit2_is_pruned = HashMap::from([(Arc::from("wit1"), Value::u1(0))]);
+        assert_finalize_ok::<Core>(s, &wit2_is_pruned, &());
+
+        let wit2_is_missing = HashMap::from([(Arc::from("wit1"), Value::u1(1))]);
+        // FIXME The finalization should fail
+        // This doesn't happen because we do bottom-up pruning:
+        // Unpopulated witness nodes are pruned, regardless of whether they will be executed
+        // Top-down pruning would execute the program and throw an error at unpopulated witness nodes
```

Comment:
In 60efcbf0a34204b28bc063804df609f523374c46:

I think this comment is confusingly worded. Really the problem is that we don't execute during finalization, so we can't determine in general exactly which nodes need to be pruned. "Top-down" vs "bottom-up" is beside the point.

My original plan was to just add an execution pass at the end of `finalize()` with some extra context passed into the execution engine to track which nodes should be pruned. Maybe you have a more efficient idea?

Reply:
Running the Bit Machine at the end of the current finalization method would work.

In the future, I think we can drop the existing pruning and retyping, to simplify the algorithm. Bit Machine execution is necessary in general and sufficient to detect all unused branches.

We finalize the types of the shared program and run the Bit Machine to detect unused branches. We might need to re-run the program because types changed. TBH, I am still confused how much the types can change in each pruning iteration and what the worst case is, but the execution-only algorithm should work.

Reply:
We shouldn't need to re-run the program. After pruning we should find that some times have shrunk, but only to eliminate entirely unused branches of types.

One complication is that we need to parse witnessess knowing that the user might provide a larger witness than is needed. For example, you could imagine a witness node that could take a type `Prod(X, Y)` but only one of the `X` or `Y`s are actually used depending on some other witness bit. Ideally we would be ok if the user actually provided both, and as part of our pruning we would shrink the witness. (The case where the user wants to provide only one, because he knows which one will actually be used, is a bit trickier. Initially I think we will need to just request he fill out the "whole type" perhaps by adding dummy data.)

But we could add a `debug_assert` or something that re-running the program succeeds and does not result in any further pruning.